File: momentra/momentra/Shell/PersonalEmpty/PersonalTabLoad.swift
import Foundation

enum PersonalTabLoad {
    static let pulseActivityLimit = 5

    /// Parallel pulse + activity fetch with in-memory SWR cache write.
    static func loadPulseTab(momentId: String?) async throws -> PersonalPulseTabData {
        let mark = ShellPerf.start("pulse_tab_ready")
        async let pulseTask = APIClient.shared.getPersonalPulse(momentId: momentId)
        async let activityTask = APIClient.shared.listPersonalActivity(momentId: momentId, limit: pulseActivityLimit)
        let pulse = try await pulseTask
        let activities = try await activityTask
        PersonalTabDataCache.putPulse(momentId: momentId, pulse: pulse, activities: activities)
        ShellPerf.end(mark, extras: ["context": "PERSONAL", "parallel": true])
        return PersonalPulseTabData(pulse: pulse, activities: activities)
    }
}

File: apk/app/src/main/java/com/example/momentra/ui/shell/personal/shared/PersonalTabLoad.kt
package com.example.momentra.ui.shell.personal.shared

import com.example.momentra.data.api.ActivityItemDto
import com.example.momentra.data.api.PersonalPulseDto
import com.example.momentra.data.repository.PersonalSliceRepository
import com.example.momentra.ui.shell.perf.ShellPerf
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope

/** Pulse-tab preview: 5 activity rows keeps Personal Pulse snappy (full list uses dedicated flow). */
const val PERSONAL_PULSE_ACTIVITY_LIMIT = 5

/**
 * Parallel pulse + activity fetch with in-memory SWR cache.
 * Callers should paint [PersonalTabDataCache.peek] before awaiting this when possible.
 */
suspend fun loadPersonalPulseTab(
    repository: PersonalSliceRepository,
    momentId: String?,
    activityLimit: Int = PERSONAL_PULSE_ACTIVITY_LIMIT,
): Result<Pair<PersonalPulseDto, List<ActivityItemDto>>> = runCatching {
    val mark = ShellPerf.start("pulse_tab_ready")
    coroutineScope {
        val pulseDeferred = async { repository.getPulse(momentId = momentId) }
        val activityDeferred = async {
            repository.getActivity(momentId = momentId, limit = activityLimit)
        }
        val pulse = pulseDeferred.await().getOrThrow()
        val activities = activityDeferred.await().getOrThrow().items
        PersonalTabDataCache.put(momentId, pulse, activities)
        ShellPerf.end(mark, mapOf("context" to "PERSONAL", "parallel" to true, "cached" to false))
        pulse to activities
    }
}
